File: examples/pong/pong_unity/PaddleController.cs
using UnityEngine;

// Unity APIs used: MonoBehaviour, Rigidbody2D, Input.GetAxis, Vector2, Time.fixedDeltaTime
public class PaddleController : MonoBehaviour
{
    [SerializeField] private float speed = 10f;
    [SerializeField] private float boundY = 4f;
    [SerializeField] private string inputAxis = "Vertical";

    private Rigidbody2D rb;

    void Start()
    {
        rb = GetComponent<Rigidbody2D>();
    }

    void FixedUpdate()
    {
        float input = Input.GetAxis(inputAxis);
        Vector2 velocity = new Vector2(0, input * speed);
        rb.velocity = velocity;

        // Clamp position
        Vector2 pos = transform.position;
        if (pos.y > boundY)
        {
            transform.position = new Vector2(pos.x, boundY);
            rb.velocity = Vector2.zero;
        }
        else if (pos.y < -boundY)
        {
            transform.position = new Vector2(pos.x, -boundY);
            rb.velocity = Vector2.zero;
        }
    }
}

File: examples/pong/pong_python/paddle_controller.py
"""Paddle controller — Python equivalent of PaddleController.cs"""

from src.engine.core import MonoBehaviour
from src.engine.math.vector import Vector2
from src.engine.input_manager import Input
from src.engine.physics.rigidbody import Rigidbody2D


class PaddleController(MonoBehaviour):

    def __init__(self):
        super().__init__()
        self.speed: float = 10.0
        self.bound_y: float = 4.0
        self.input_axis: str = "Vertical"

    def start(self):
        self.rb = self.get_component(Rigidbody2D)

    def fixed_update(self):
        input_val = Input.get_axis(self.input_axis)
        velocity = Vector2(0, input_val * self.speed)
        self.rb.velocity = velocity

        # Clamp position
        pos = self.transform.position
        if pos.y > self.bound_y:
            self.transform.position = Vector2(pos.x, self.bound_y)
            self.rb.velocity = Vector2(0, 0)
        elif pos.y < -self.bound_y:
            self.transform.position = Vector2(pos.x, -self.bound_y)
            self.rb.velocity = Vector2(0, 0)
